applets/notification_area/na-grid.c

```diff
@@ -84,17 +98,157 @@ compare_items (gconstpointer a,
+      row = data->index % data->na_grid->grid_props->rows;
+    } else {
+      row = data->index / data->na_grid->grid_props->cols;
+      col = data->index % data->na_grid->grid_props->cols;
+    }
+
+  /* only update item position if has changed from current */
+  gtk_container_child_get (GTK_CONTAINER (data->na_grid),
+                           item,
+                           "left-attach", &left,
+                           "width", &right,
+                           "top-attach", &top,
+                           "height", &bottom,
+                           NULL);
+  if (left != col || right != 1 || top != row || bottom != 1)
+    {
+      gtk_container_child_set (GTK_CONTAINER (data->na_grid),
+                               item,
+                               "left-attach", col,
+                               "width", 1,
+                               "top-attach", row,
+                               "height", 1,
+                               NULL);
+    }
+  /* increment to index of next item */
+  data->index++;
+}
+
+static void
+determine_grid_properties (NaGrid *self)
+{
+  gint rows, cols, length;
+  GtkOrientation orientation;
+  GtkAllocation allocation;
+  
+  orientation = gtk_orientable_get_orientation (GTK_ORIENTABLE (self));
+  gtk_widget_get_allocation (GTK_WIDGET (self), &allocation);
+  length = g_slist_length (self->items);
+
+  if (orientation == GTK_ORIENTATION_HORIZONTAL)
+    {
+      rows = MAX (1, allocation.height / self->min_icon_size);
+      cols = MAX (1, length / rows);
+      if (length % rows)
+        cols++;
+    } else {
+      cols = MAX (1, allocation.width / self->min_icon_size);
+      rows = MAX (1, length / cols);
+      if (length % cols)
+        rows++;
+    }
+  
+  self->grid_props->cols = cols;
+  self->grid_props->rows = rows;
+  self->grid_props->length = length;
+}
+
 static void
-reorder_items (GtkWidget *widget,
-               gpointer   user_data)
+refresh_grid (NaGrid *self)
 {
-  NaGrid *nb;
-  gint position;
+  PackData pack_data;
+  GSList *item;
+  GtkOrientation orientation;
+  gint row = 0, col = 0;
+  
+  g_return_if_fail (NA_IS_GRID (self));
+  
+  orientation = gtk_orientable_get_orientation (GTK_ORIENTABLE (self));
+  determine_grid_properties (self);
+
+  /*
+  * There might be a better way to do this
+  * The refresh is caused by all items being removed from the container first,
+  * then re-added at the position they're supposed to be at,
+  * then sorting them again in reorder_items_with_data()
+  *
+  * I couldn't find how to force the update with another mechanism?
+  * Just sorting them doesn't work.
+  * Removing them/re-adding them alone doesn't cause an update either.
+  * All these steps have to happen for the grid to update when the panel loading occurs, when a new item is added or when the min icon size is changed in the preferences...
+  */
+  for (item = self->items; item; item = item->next)
+    {
+      if (GTK_IS_WIDGET (item))
```

Comment:
Can `items` list contain something that is not `GtkWidget`? If not then this check should not be needed...